### DISCUSSION TITLE
Creating a custom environment

### DISCUSSION THREAD

**Discussion by Yabbanootje** [PKU-Alignment/safety-gymnasium#94] [Q&A]
I want to create my own environment, where I want hazards to be in specific locations. I have created a class that inherits `BaseTask`, just like the example of `GoalLevel0` on the documentation page. I then register this class using the `register()` function. The next thing I do is `make()` an environment using the id I gave when registering. However, I get the error that I do not give a `config` to the task. But I do not know where I should get this config from, especially because I do not need to give a config if I want to use `SafetyPointGoal0-v0` in the same way for example. I have the feeling that I am missing some steps, but I cannot find which ones. Could someone help me with this?

**Comment by muchvo** [comment]
Hello. To assist you more effectively, could you please provide the exact error message you're encountering, along with a minimal reproducible example? This would greatly help us in identifying the root cause of the issue.

In the meantime, you might want to try adding the following line of code to your `__init__` method:

```python
self._is_load_static_geoms = False
```

Please let us know if this helps resolve the issue or if you encounter any further problems. Your feedback is valuable to improve the project.

**Reply by Yabbanootje** [reply]
Hello. The exact error message is `TypeError: CurriculumLevel0.__init__() missing 1 required positional argument: 'config'`.
My code looks something like the following.

I have a class that tries to place apples in specific locations:
```
class CurriculumLevel0(BaseTask):
    def __init__(self, config):
        super().__init__(config=config)
        self._is_load_static_geoms = False
        self._add_geoms(Apples(num=2, size=0.3, placements=[(-1, -1), (1, 1)]))

    def calculate_reward(self):
        pass

    def specific_reset(self):
        pass

    def specific_step(self):
        pass

    def update_world(self):
        pass
```
I added the line you suggested, but it did not help with the error.

In the same directory (custom_envs) I have an `__init__.py` file with:
```
register(id="Curriculum0-v0", entry_point="custom_envs.curriculum_level_0:CurriculumLevel0")
```

Then in another python file in the same directory I run the following:
```
env = safety_gymnasium.vector.make(env_id="Curriculum0-v0")
```

This is also the line where I get the error. However, if I replace the id with one that is standard in safety gymnasium, I do not get the same error. Do you know what I am doing wrong?

**Reply by muchvo** [reply]
Thank you for bringing this up. I understand the issue now. Indeed, we have provided a configuration system that allows different tasks to reuse some common code. The configurations for tasks are statically defined. For registering your environment, please refer to [code here](https://github1s.com/PKU-Alignment/safety-gymnasium/blob/HEAD/safety_gymnasium/__init__.py). If you need further assistance, feel free to ask!

**Reply by Yabbanootje** [reply]
Thank you for your quick response. I have come further, but I have reached a new problem. I think that I have to register my task somehow, but I cannot find how. 

I get the error: `AssertionError: Task=CurriculumLevel0 not implemented.`

My `__init__.py` file has changed to:
```
from custom_envs.curriculum_level_0 import CurriculumLevel0

def __register_helper(env_id, entry_point, spec_kwargs=None, **kwargs):
    ...

def __combine(tasks, agents, max_episode_steps):
    ...

goal_tasks = {'Curriculum0': {}, 'Curriculum1': {}, 'Curriculum2': {}}
__combine(goal_tasks, robots, max_episode_steps=1000)
```
Where both functions are copied directly from the code you sent and I omitted the variables I also copied.

When calling the slightly updated `make()` function as:
```
env = safety_gymnasium.vector.make(env_id='SafetyPointCurriculum0-v0')
```
I get the error above. Could you also help me with this issue?

**Reply by muchvo** [reply]
You might need to import your class into [this file](https://github1s.com/PKU-Alignment/safety-gymnasium/blob/HEAD/safety_gymnasium/tasks/__init__.py).

**Reply by Yabbanootje** [reply]
Is there any way to do that without changing your source code?

**Reply by muchvo** [reply]
I apologize, but the approach you mentioned isn't feasible. Based on your question, I'm assuming you've installed the package using pip, which would make it inconvenient for you to modify the file I pointed out. As an alternative, you can download the source code from GitHub and then install it in editable mode using the command `pip install -e .` from the root directory of the project. This will allow you to make the suggested modifications more easily.

**Reply by Yabbanootje** [reply]
I am running my code through docker, so it is a bit more difficult for me to do, but I suppose that is the next issue I should fix. 
But is there really no way to make a static environment with your code? My idea is to make a curriculum of tasks, where I want the obstacles to be at the edge of the area and come closer to the goal with each new task. Is there maybe a way to get the seeds of the environments I want so that I can use your environments?
In any case thank you for your answers!

**Reply by muchvo** [reply]
You can retrieve the seed by accessing the `env._seed` attribute. For more details, please refer to the code provided [here in 157](https://github1s.com/PKU-Alignment/safety-gymnasium/blob/HEAD/safety_gymnasium/builder.py). Furthermore, if you need to have obstacles positioned at the edges, you should adjust the `placements`, `locations`, and `keepout` properties according to your requirements. You can find them in this snippet of code [(33-35)](https://github1s.com/PKU-Alignment/safety-gymnasium/blob/HEAD/safety_gymnasium/assets/free_geoms/vases.py). Currently, the feature to have obstacles of every new task closer towards the goal is not supported within the same task object due to static configuring.

**Reply by Yabbanootje** [reply]
Okay that helps! Thank you very much for your assistance!

**Reply by muchvo** [reply]
Additionally, while a single task object cannot change the area in which random objects are generated, you can achieve this by creating new instances of the task object at different training stages. I've also taken note of your mention of a `static` environment; if you're referring to a non-randomized task setup, you have the option to disable the randomness. Please refer to this discussion [discussion](https://github.com/PKU-Alignment/safety-gymnasium/discussions/87) for more details.

**Comment by muchvo** [comment]
> Is there any way to do that without changing your source code?

Someone has suggested [a method](https://github.com/PKU-Alignment/safety-gymnasium/issues/96) that allows registering new tasks without modifying the source files.
